### PR TITLE
Grant API permissions programmatically

### PR DESCRIPTION
**Pull request by msewaweru**

### CODE REVIEW

**Comment by msewaweru** [comment]
@Jackson-Woods , @psignoret, requesting for your review on this PR. Thanks

**Comment by msewaweru** [comment]
@Jackson-Woods , @psignoret polite reminder to review and approve this work. All review comments added to the [Graph API PR](https://github.com/microsoftgraph/microsoft-graph-docs/pull/16217) have been incorporated. Thanks.

File: microsoftgraph/docs-conceptual/tutorial-grant-app-only-api-permissions.md [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+
+Before you can grant app roles, you must first identify the app roles to grant and the resource service principal that exposes the app roles. App roles are defined in the `appRoles` object of a service principal. 
+
+In this article, you'll use the `Microsoft Graph` service principal in the tenant as your resource service principal.
+
+```powershell
+Get-MgServicePrincipal -Filter "displayName eq 'Microsoft Graph'" -Property AppRoles | Select -ExpandProperty appRoles |fl
+```
+
+```Output
```
Comment (Jackson-Woods) [inline microsoftgraph/docs-conceptual/tutorial-grant-app-only-api-permissions.md:48]: Do we need to state that the output is truncated?

File: microsoftgraph/docs-conceptual/tutorial-grant-app-only-api-permissions.md [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+Id                   : df021288-bdef-4463-88db-98f22de89214
+IsEnabled            : True
+Origin               : Application
+Value                : User.Read.All
+AdditionalProperties : {}
+```
+
+## Step 2: Create a client service principal
+
+The first step in granting consent is to [create the service principal](/powershell/module/microsoft.graph.applications/new-mgserviceprincipal?view=graph-powershell-1.0&preserve-view=true). To do so, you'll need the `App Id` of your application.
```
Comment (Jackson-Woods) [inline microsoftgraph/docs-conceptual/tutorial-grant-app-only-api-permissions.md:70]: I would slightly expand this - "[create a service principal] for the app that you'll grant permissions."

File: microsoftgraph/docs-conceptual/tutorial-grant-app-only-api-permissions.md [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+DisplayName    : My application
+AppId          : 05210c44-437f-4a40-bd38-b5b4eaf251ef
+SignInAudience : AzureADandPersonalMicrosoftAccount
+```
+
+## Step 3: Assign an app role to the client service principal
+
+In this step, you'll assign an app role exposed by your resource app to the service principal we created in step 2. To create an app role assignment, you'll need the following information:
+
+1. **PrincipalId** - object Id of the service principal authorized for direct access.
```
Comment (Jackson-Woods) [inline microsoftgraph/docs-conceptual/tutorial-grant-app-only-api-permissions.md:107]: **PrincipalId** - object Id of the service principal *to be* authorized for direct access.? Just a thought.

File: microsoftgraph/docs-conceptual/tutorial-grant-delegated-api-permissions.md [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+
+1. A working Azure AD tenant.
+1. Microsoft Graph PowerShell SDK is installed. Follow the [Install the Microsoft Graph PowerShell SDK](installation.md) guide to install the SDK.
+1. Microsoft Graph PowerShell using a global administrator role and the appropriate permissions. For this tutorial, the `Application.Read.All` and `DelegatedPermissionGrant.ReadWrite.All` delegated permissions are required. To set the permissions in Microsoft Graph PowerShell, run;
+
+    ```powershell
+    Connect-MgGraph -Scopes "Application.ReadWrite.All", "DelegatedPermissionGrant.ReadWrite.All"
+    ```
+
+    Select **Consent on behalf of your organization** before accepting in the login dialog box.
```
Comment (Jackson-Woods) [inline microsoftgraph/docs-conceptual/tutorial-grant-delegated-api-permissions.md:33]: Same - why consent for the entire org?

File: microsoftgraph/docs-conceptual/tutorial-grant-delegated-api-permissions.md [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+
+In this tutorial, you'll grant delegated permissions that are exposed by an API to an app. Delegated permissions, also called scopes or OAuth2 permissions, allow an app to call an API on behalf of a signed-in user.
+
+## Prerequisites
+
+To successfully complete this tutorial, make sure you have the required prerequisites:
+
+1. A working Azure AD tenant.
+1. Microsoft Graph PowerShell SDK is installed. Follow the [Install the Microsoft Graph PowerShell SDK](installation.md) guide to install the SDK.
+1. Microsoft Graph PowerShell using a global administrator role and the appropriate permissions. For this tutorial, the `Application.Read.All` and `DelegatedPermissionGrant.ReadWrite.All` delegated permissions are required. To set the permissions in Microsoft Graph PowerShell, run;
```
Comment (Jackson-Woods) [inline microsoftgraph/docs-conceptual/tutorial-grant-delegated-api-permissions.md:27]: Same - ";" vs ":" at end of the line.

File: microsoftgraph/docs-conceptual/tutorial-grant-delegated-api-permissions.md [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+UserConsentDescription  : Allows the app to list groups, and to read their properties and all group memberships on your behalf.  Also allows the app to read calendar, conversations, files, and other group
+                          content for all groups you can access.
+UserConsentDisplayName  : Read all groups
+Value                   : Group.Read.All
+AdditionalProperties    : {}
+```
+
+## Step 2: Create a client service principal
+
+The first step in granting consent is to [create the service principal](/powershell/module/microsoft.graph.applications/new-mgserviceprincipal?view=graph-powershell-1.0&preserve-view=true). To do so, you'll need the `App Id` of your application.
```
Comment (Jackson-Woods) [inline microsoftgraph/docs-conceptual/tutorial-grant-delegated-api-permissions.md:76]: Same - "[create a service principal] for the app that you'll grant permissions."

File: microsoftgraph/docs-conceptual/tutorial-grant-delegated-api-permissions.md [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+ClientId    : 22c1770d-30df-49e7-a763-f39d2ef9b369
+ConsentType : AllPrincipals
+PrincipalId :
+ResourceId  : a67ad0d0-a7d1-4adb-8cd9-bcdd0c866d3c
+Scope       : Group.Read.All
+```
+
+### Step 4: Assign more or revoke delegated permissions to the service principal
+
+You can add more or reduce scopes to an already existing oauth2PermissionGrant object.
```
Comment (Jackson-Woods) [inline microsoftgraph/docs-conceptual/tutorial-grant-delegated-api-permissions.md:144]: Maybe a little more direct? "You can add or remove scopes from an existing oauth2PermissionGrant object."

File: microsoftgraph/docs-conceptual/tutorial-grant-delegated-api-permissions.md [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+SignInAudience : AzureADandPersonalMicrosoftAccount
+```
+
+## Step 3: Grant delegated permission to the client service principal
+
+To create a delegated permission grant, you'll need the following information:
+
+1. **ClientId** - object Id of the client service principal authorized to act on behalf of the user. In this case, the service principal we created in step 2.
+1. **ConsentType** - `AllPrincipals` to authorize all users in the tenant or `Principal` for a single user.
+1. **PrincipalId** - Id of the user for *Principal* consents, `null` for *AllPrincipals* consents.
```
Comment (Jackson-Woods) [inline microsoftgraph/docs-conceptual/tutorial-grant-delegated-api-permissions.md:115]: Suggest reversing the order of this line to match the line right above - AllPrincipals, then Principal.

File: microsoftgraph/docs-conceptual/tutorial-grant-delegated-api-permissions.md [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+1. **ResourceId** - object Id of the service principal representing the resource app in the tenant.
+1. **Scope** - space-delimited list of permission claim values, for example `User.Read.All`.
+
+In this example, the object Id of the resource service principal is `2cab1707-656d-40cc-8522-3178a184e03d`. You'll grant `Group.Read.All` scope to the service principal and grant consent on behalf of all users in the tenant.
```
Comment (Jackson-Woods) [inline microsoftgraph/docs-conceptual/tutorial-grant-delegated-api-permissions.md:119]: "You'll grant *the* `Group.Read.All` scope..."

**Review by Jackson-Woods — APPROVED** [review]
Looks good Eunice, thanks! See minor comments on the docs.

File: microsoftgraph/docs-conceptual/tutorial-grant-app-only-api-permissions.md [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+
+1. A working Azure AD tenant.
+1. Microsoft Graph PowerShell SDK is installed. Follow the [Install the Microsoft Graph PowerShell SDK](installation.md) guide to install the SDK.
+1. Microsoft Graph PowerShell using a global administrator role and the appropriate permissions. For this tutorial, the `Application.Read.All` and `AppRoleAssignment.ReadWrite.All` delegated permissions are required. To set the permissions in Microsoft Graph PowerShell, run;
+
+    ```powershell
+    Connect-MgGraph -Scopes "Application.ReadWrite.All", "AppRoleAssignment.ReadWrite.All"
+    ```
+
+    Select **Consent on behalf of your organization** before accepting in the login dialog box.
```
[the remark this answers is not in the document]
Comment (msewaweru) [reply microsoftgraph/docs-conceptual/tutorial-grant-app-only-api-permissions.md]: > Why should they consent for the entire org?

Removed that line. Thanks.

**Comment by msewaweru** [comment]
> Looks good Eunice, thanks! See minor comments on the docs. Please do address the point about consenting for the entire org, which is potentially a significant overprivilege situation.

@Jackson-Woods , thank you for your comprehensive feedback. I have acted on all your feedback and will be proceeding to merge. Thanks.

**Review by timayabi2020 — APPROVED** [review]

**Pull request merged**